File: tests/algorithms/x/Java/machine_learning/data_transformations.java
public class Main {

    static double floor(double x) {
        long i = ((Number)(x)).intValue();
        if ((((Number)(i)).doubleValue()) > x) {
            i = i - 1;
        }
        return ((Number)(i)).doubleValue();
    }

    static double pow10(long n) {
        double result = 1.0;
        long i_2 = 0L;
        while (i_2 < n) {
            result = result * 10.0;
            i_2 = i_2 + 1;
        }
        return result;
    }

    static double round(double x, long n) {
        double m = pow10(n);
        double y_1 = ((Number)(floor(x * m + 0.5))).doubleValue();
        return y_1 / m;
    }

    static double sqrtApprox(double x) {
        double guess = x;
        long i_4 = 0L;
        while (i_4 < 20) {
            guess = (guess + x / guess) / 2.0;
            i_4 = i_4 + 1;
        }
        return guess;
    }

    static double mean(double[] data) {
        double total = 0.0;
        long i_6 = 0L;
        long n_1 = data.length;
        while (i_6 < n_1) {
            total = total + data[(int)(i_6)];
            i_6 = i_6 + 1;
        }
        return total / (((Number)(n_1)).doubleValue());
    }

    static double stdev(double[] data) {
        long n_2 = data.length;
        if (n_2 <= 1) {
            throw new RuntimeException(String.valueOf("data length must be > 1"));
        }
        double m_2 = mean(((double[])(data)));
        double sum_sq_1 = 0.0;
        long i_8 = 0L;
        while (i_8 < n_2) {
            double diff_1 = data[(int)(i_8)] - m_2;
            sum_sq_1 = sum_sq_1 + diff_1 * diff_1;
            i_8 = i_8 + 1;
        }
        return sqrtApprox(sum_sq_1 / (((Number)((n_2 - 1))).doubleValue()));
    }

    static double[] normalization(double[] data, long ndigits) {
        double x_min = ((Number)(_min(data))).doubleValue();
        double x_max_1 = ((Number)(_max(data))).doubleValue();
        double denom_1 = x_max_1 - x_min;
        double[] result_2 = ((double[])(new double[]{}));
        long i_10 = 0L;
        long n_4 = data.length;
        while (i_10 < n_4) {
            double norm_1 = (data[(int)(i_10)] - x_min) / denom_1;
            result_2 = ((double[])(appendDouble(result_2, round(norm_1, ndigits))));
            i_10 = i_10 + 1;
        }
        return result_2;
    }

    static double[] standardization(double[] data, long ndigits) {
        double mu = mean(((double[])(data)));
        double sigma_1 = stdev(((double[])(data)));
        double[] result_4 = ((double[])(new double[]{}));
        long i_12 = 0L;
        long n_6 = data.length;
        while (i_12 < n_6) {
            double z_1 = (data[(int)(i_12)] - mu) / sigma_1;
            result_4 = ((double[])(appendDouble(result_4, round(z_1, ndigits))));
            i_12 = i_12 + 1;
        }
        return result_4;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            System.out.println(_p(normalization(((double[])(new double[]{2.0, 7.0, 10.0, 20.0, 30.0, 50.0})), 3L)));
            System.out.println(_p(normalization(((double[])(new double[]{5.0, 10.0, 15.0, 20.0, 25.0})), 3L)));
            System.out.println(_p(standardization(((double[])(new double[]{2.0, 7.0, 10.0, 20.0, 30.0, 50.0})), 3L)));
            System.out.println(_p(standardization(((double[])(new double[]{5.0, 10.0, 15.0, 20.0, 25.0})), 3L)));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static double[] appendDouble(double[] arr, double v) {
        double[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static Object[] _toObjectArray(Object v) {
        if (v instanceof Object[]) return (Object[]) v;
        if (v instanceof int[]) return java.util.Arrays.stream((int[]) v).boxed().toArray();
        if (v instanceof double[]) return java.util.Arrays.stream((double[]) v).boxed().toArray();
        if (v instanceof long[]) return java.util.Arrays.stream((long[]) v).boxed().toArray();
        if (v instanceof boolean[]) { boolean[] a = (boolean[]) v; Object[] out = new Object[a.length]; for (int i = 0; i < a.length; i++) out[i] = a[i]; return out; }
        return (Object[]) v;
    }

    static double _min(Object a) {
        Object[] arr = _toObjectArray(a);
        double m = ((Number)arr[0]).doubleValue();
        for (int i = 1; i < arr.length; i++) {
            double v = ((Number)arr[i]).doubleValue();
            if (v < m) m = v;
        }
        return m;
    }

    static double _max(Object a) {
        Object[] arr = _toObjectArray(a);
        double m = ((Number)arr[0]).doubleValue();
        for (int i = 1; i < arr.length; i++) {
            double v = ((Number)arr[i]).doubleValue();
            if (v > m) m = v;
        }
        return m;
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            if (d == Math.rint(d)) return String.valueOf((long) d);
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/machine_learning/data_transformations.pas
{$mode objfpc}
program Main;
uses SysUtils, Math;
type RealArray = array of real;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
function min_real(xs: array of real): real;
var i: integer; m: real;
begin
  if Length(xs) = 0 then begin min_real := 0; exit; end;
  m := xs[0];
  for i := 1 to High(xs) do if xs[i] < m then m := xs[i];
  min_real := m;
end;
function max_real(xs: array of real): real;
var i: integer; m: real;
begin
  if Length(xs) = 0 then begin max_real := 0; exit; end;
  m := xs[0];
  for i := 1 to High(xs) do if xs[i] > m then m := xs[i];
  max_real := m;
end;
function list_real_to_str(xs: array of real): string;
var i: integer;
begin
  Result := '[';
  for i := 0 to High(xs) do begin
    Result := Result + FloatToStr(xs[i]);
    if i < High(xs) then Result := Result + ' ';
  end;
  Result := Result + ']';
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  data: RealArray;
  x: real;
  n: integer;
  ndigits: integer;
function floor(x: real): real; forward;
function pow10(n: integer): real; forward;
function round(x: real; n: integer): real; forward;
function sqrtApprox(x: real): real; forward;
function mean(data: RealArray): real; forward;
function stdev(data: RealArray): real; forward;
function normalization(data: RealArray; ndigits: integer): RealArray; forward;
function standardization(data: RealArray; ndigits: integer): RealArray; forward;
function floor(x: real): real;
var
  floor_i: integer;
begin
  floor_i := Trunc(x);
  if Double(floor_i) > x then begin
  floor_i := floor_i - 1;
end;
  exit(Double(floor_i));
end;
function pow10(n: integer): real;
var
  pow10_result_: real;
  pow10_i: integer;
begin
  pow10_result_ := 1;
  pow10_i := 0;
  while pow10_i < n do begin
  pow10_result_ := pow10_result_ * 10;
  pow10_i := pow10_i + 1;
end;
  exit(pow10_result_);
end;
function round(x: real; n: integer): real;
var
  round_m: real;
  round_y: real;
begin
  round_m := pow10(n);
  round_y := Double(Floor((x * round_m) + 0.5));
  exit(round_y / round_m);
end;
function sqrtApprox(x: real): real;
var
  sqrtApprox_guess: real;
  sqrtApprox_i: integer;
begin
  sqrtApprox_guess := x;
  sqrtApprox_i := 0;
  while sqrtApprox_i < 20 do begin
  sqrtApprox_guess := (sqrtApprox_guess + (x / sqrtApprox_guess)) / 2;
  sqrtApprox_i := sqrtApprox_i + 1;
end;
  exit(sqrtApprox_guess);
end;
function mean(data: RealArray): real;
var
  mean_total: real;
  mean_i: integer;
  mean_n: integer;
begin
  mean_total := 0;
  mean_i := 0;
  mean_n := Length(data);
  while mean_i < mean_n do begin
  mean_total := mean_total + data[mean_i];
  mean_i := mean_i + 1;
end;
  exit(mean_total / Double(mean_n));
end;
function stdev(data: RealArray): real;
var
  stdev_n: integer;
  stdev_m: real;
  stdev_sum_sq: real;
  stdev_i: integer;
  stdev_diff: real;
begin
  stdev_n := Length(data);
  if stdev_n <= 1 then begin
  panic('data length must be > 1');
end;
  stdev_m := mean(data);
  stdev_sum_sq := 0;
  stdev_i := 0;
  while stdev_i < stdev_n do begin
  stdev_diff := data[stdev_i] - stdev_m;
  stdev_sum_sq := stdev_sum_sq + (stdev_diff * stdev_diff);
  stdev_i := stdev_i + 1;
end;
  exit(sqrtApprox(stdev_sum_sq / Double(stdev_n - 1)));
end;
function normalization(data: RealArray; ndigits: integer): RealArray;
var
  normalization_x_min: real;
  normalization_x_max: real;
  normalization_denom: real;
  normalization_result_: array of real;
  normalization_i: integer;
  normalization_n: integer;
  normalization_norm: real;
begin
  normalization_x_min := Double(min_real(data));
  normalization_x_max := Double(max_real(data));
  normalization_denom := normalization_x_max - normalization_x_min;
  normalization_result_ := [];
  normalization_i := 0;
  normalization_n := Length(data);
  while normalization_i < normalization_n do begin
  normalization_norm := (data[normalization_i] - normalization_x_min) / normalization_denom;
  normalization_result_ := concat(normalization_result_, [round(normalization_norm, ndigits)]);
  normalization_i := normalization_i + 1;
end;
  exit(normalization_result_);
end;
function standardization(data: RealArray; ndigits: integer): RealArray;
var
  standardization_mu: real;
  standardization_sigma: real;
  standardization_result_: array of real;
  standardization_i: integer;
  standardization_n: integer;
  standardization_z: real;
begin
  standardization_mu := mean(data);
  standardization_sigma := stdev(data);
  standardization_result_ := [];
  standardization_i := 0;
  standardization_n := Length(data);
  while standardization_i < standardization_n do begin
  standardization_z := (data[standardization_i] - standardization_mu) / standardization_sigma;
  standardization_result_ := concat(standardization_result_, [round(standardization_z, ndigits)]);
  standardization_i := standardization_i + 1;
end;
  exit(standardization_result_);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(list_real_to_str(normalization([2, 7, 10, 20, 30, 50], 3)));
  writeln(list_real_to_str(normalization([5, 10, 15, 20, 25], 3)));
  writeln(list_real_to_str(standardization([2, 7, 10, 20, 30, 50], 3)));
  writeln(list_real_to_str(standardization([5, 10, 15, 20, 25], 3)));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
